core/src/main/java/org/elasticsearch/indices/IndicesModule.java

```diff
@@ -147,11 +169,33 @@ private void registerBuiltinWritables() {
             }
         }
 
-        // we register _field_names here so that it has a chance to see all other mappers, including from plugins
-        metadataMappers.put(FieldNamesFieldMapper.NAME, new FieldNamesFieldMapper.TypeParser());
+        // we register _field_names here so that it has a chance to see all the other mappers, including from plugins
+        metadataMappers.put(fieldNamesEntry.getKey(), fieldNamesEntry.getValue());
         return Collections.unmodifiableMap(metadataMappers);
     }
 
+    public static Set<String> getBuiltInMetaDataFields() {
+        return builtInMetadataMappers.keySet();
+    }
+
+    private static BiPredicate<String, String> getFieldFilter(List<MapperPlugin> mapperPlugins) {
+        BiPredicate<String, String> fieldFilter = MapperPlugin.NOOP_FIELD_FILTER;
+        for (MapperPlugin mapperPlugin : mapperPlugins) {
+            fieldFilter = and(fieldFilter, mapperPlugin.getFieldFilter());
```

Comment:
Why not `fieldFilter.and(mapperPlugin.getFieldFilter()`.

Reply:
Ah. One nice thing about the way you have it now is that we don't end up with a lot of NOOPs all chained together when there are many mapper plugins. Which feels worth it to me. Maybe just document that that is the point of `and`?

Reply:
ok yea the point is that we do not do anything in this case, otherwise we parse the mappings and go through them for nothing.